### ISSUE TITLE
isort behavior different on linux vs osx even with config

### ISSUE THREAD

**Issue by ahonnecke** [timothycrosley/isort#725]
on OSX:
```bash
web on  fox-295 [⇡$!] via 🐍 3.6.4 took 2s
➜ pipenv run isort --recursive --check-only server
ERROR: /Users/ahonnecke/Code/repos/web/server/portfolio_manager/graphql/tests/test_ledger_addresses_query.py Imports are incorrectly sorted.
ERROR: /Users/ahonnecke/Code/repos/web/server/portfolio_manager/wallets/test_supported_ledgers.py Imports are incorrectly sorted.


web on  fox-295 [⇡$!] via 🐍 3.6.4
➜ head -n16 /Users/ahonnecke/Code/repos/web/server/portfolio_manager/graphql/tests/test_ledger_addresses_query.py
import json
import os
from time import sleep
from unittest.mock import MagicMock

from django.test import TestCase

from wiremock.constants import Config
from wiremock.resources.mappings import Mapping
from wiremock.resources.mappings.resource import Mappings
from wiremock.server import WireMockServer


from assets.models import Ledger
from dad.configuration import PortfolioManagerConfiguration
from portfolio_manager.models import LedgerAddress, Portfolio

web on  fox-295 [⇡$!] via 🐍 3.6.4
➜ head -n16 /Users/ahonnecke/Code/repos/web/server/portfolio_manager/wallets/test_supported_ledgers.py
import json
import os
from time import sleep
from unittest.mock import MagicMock

from django.test import TestCase

from wiremock.constants import Config
from wiremock.resources.mappings import Mapping
from wiremock.resources.mappings.resource import Mappings
from wiremock.server import WireMockServer


from assets.models import Ledger
from dad.configuration import PortfolioManagerConfiguration
from portfolio_manager.wallets.ledger_client import LedgerClient

web on  fox-295 [⇡$!] via 🐍 3.6.4
➜ pipenv run isort --recursive --check-only server
ERROR: /Users/ahonnecke/Code/repos/web/server/portfolio_manager/graphql/tests/test_ledger_addresses_query.py Imports are incorrectly sorted.
ERROR: /Users/ahonnecke/Code/repos/web/server/portfolio_manager/wallets/test_supported_ledgers.py Imports are incorrectly sorted.

web on  fox-295 [⇡$!] via 🐍 3.6.4 took 2s
➜ pipenv run isort --recursive server
Fixing /Users/ahonnecke/Code/repos/web/server/portfolio_manager/graphql/tests/test_ledger_addresses_query.py
Fixing /Users/ahonnecke/Code/repos/web/server/portfolio_manager/wallets/test_supported_ledgers.py
Skipped 26 files

web on  fox-295 [⇡$!] via 🐍 3.6.4 took 2s
➜ head -n16 /Users/ahonnecke/Code/repos/web/server/portfolio_manager/wallets/test_supported_ledgers.py
import json
import os
from time import sleep
from unittest.mock import MagicMock

from django.test import TestCase

from wiremock.constants import Config
from wiremock.resources.mappings import Mapping
from wiremock.resources.mappings.resource import Mappings
from wiremock.server import WireMockServer

from assets.models import Ledger
from dad.configuration import PortfolioManagerConfiguration
from portfolio_manager.wallets.ledger_client import LedgerClient
from util.security import create_user, get_token_for_user

web on  fox-295 [⇡$!] via 🐍 3.6.4
➜ head -n16 /Users/ahonnecke/Code/repos/web/server/portfolio_manager/graphql/tests/test_ledger_addresses_query.py
import json
import os
from time import sleep
from unittest.mock import MagicMock

from django.test import TestCase

from wiremock.constants import Config
from wiremock.resources.mappings import Mapping
from wiremock.resources.mappings.resource import Mappings
from wiremock.server import WireMockServer

from assets.models import Ledger
from dad.configuration import PortfolioManagerConfiguration
from portfolio_manager.models import LedgerAddress, Portfolio
from util.security import create_client_user, get_token_for_user
```

Cloud based CI platform (Circle CI / ubuntu)
```bash
circleci@02354f6b3c74:~/project/repos/web$ pipenv run isort --recursive --check-only server
ERROR: /home/circleci/project/repos/web/server/portfolio_manager/graphql/tests/test_ledger_addresses_query.py Imports are incorrectly sorted.
ERROR: /home/circleci/project/repos/web/server/portfolio_manager/wallets/test_supported_ledgers.py Imports are incorrectly sorted.

circleci@02354f6b3c74:~/project/repos/web$ cp /home/circleci/project/repos/web/server/portfolio_manager/graphql/tests/test_ledger_addresses_query.py /home/circleci/project/repos/web/server/portfolio_manager/graphql/tests/test_ledger_addresses_query.py.bak

circleci@02354f6b3c74:~/project/repos/web$ cp /home/circleci/project/repos/web/server/portfolio_manager/wallets/test_supported_ledgers.py /home/circleci/project/repos/web/server/portfolio_manager/wallets/test_supported_ledgers.py.bak

circleci@02354f6b3c74:~/project/repos/web$ pipenv run isort --recursive server
Fixing /home/circleci/project/repos/web/server/portfolio_manager/graphql/tests/test_ledger_addresses_query.py
Fixing /home/circleci/project/repos/web/server/portfolio_manager/wallets/test_supported_ledgers.py
Skipped 26 files
circleci@02354f6b3c74:~/project/repos/web$ diff /home/circleci/project/repos/web/server/portfolio_manager/graphql/tests/test_ledger_addresses_query.py /home/circleci/project/repos/web/server/portfolio_manager/graphql/tests/test_ledger_addresses_query.py.bak
8,12d7
< from wiremock.constants import Config
< from wiremock.resources.mappings import Mapping
< from wiremock.resources.mappings.resource import Mappings
< from wiremock.server import WireMockServer
<
17a13,16
> from wiremock.constants import Config
> from wiremock.resources.mappings import Mapping
> from wiremock.resources.mappings.resource import Mappings
> from wiremock.server import WireMockServer
circleci@02354f6b3c74:~/project/repos/web$ diff /home/circleci/project/repos/web/server/portfolio_manager/wallets/test_supported_ledgers.py /home/circleci/project/repos/web/server/portfolio_manager/wallets/test_supported_ledgers.py.bak
8,12d7
< from wiremock.constants import Config
< from wiremock.resources.mappings import Mapping
< from wiremock.resources.mappings.resource import Mappings
< from wiremock.server import WireMockServer
<
17a13,16
> from wiremock.constants import Config
> from wiremock.resources.mappings import Mapping
> from wiremock.resources.mappings.resource import Mappings
> from wiremock.server import WireMockServer
circleci@02354f6b3c74:~/project/repos/web$ pipenv run isort --recursive --check-only server
Skipped 26 files
```

isort
```
[settings]
known_django=django
sections=FUTURE,STDLIB,DJANGO,THIRDPARTY,FIRSTPARTY,LOCALFOLDER
```

TLDR, isort requires a different number of newlines on ubuntu.

**Comment by hoylemd** [comment]
I'm having a similar issue, except instead of ubuntu, I'm trying to run it in a  `python:3.6` docker container.

**Comment by ahonnecke** [comment]
Now I feel silly, I'm actually running it in a circleci/python:3.6.4-node-browsers docker container.

**Comment by hoylemd** [comment]
Interesting... Now I want to try it on a bunch of other platforms to narrow it down.  Maybe it's just docker, or the python images?

**Comment by ahonnecke** [comment]
I managed to *hackily* resolve it by ordering the imports in the isort config, so changing it to this:
```
[settings]
known_django=django
known_wiremock=wiremock
sections=FUTURE,STDLIB,DJANGO,THIRDPARTY,FIRSTPARTY,LOCALFOLDER,WIREMOCK
```

Normalized the behavior

**Comment by hoylemd** [comment]
Awesome! I'll try that on my setup tomorrow.  

I guess the problem is in how it defaults the sections order, somehow works differently depending on the platform.  Sounds like a fun bug hunt :)

**Comment by hoylemd** [comment]
Alas, this didn't work for me.  A lot of the discrepancies appear to be about the space between sections. 

Specifically, it doesn't seem to be able to tell that a local package doesn't belong in the same group as third party libraries like sqlalchemy.

(I just added the sections bit - I'm not using django or wiremock)

**Comment by ahonnecke** [comment]
I think that you need to specify the section order for each module that you're getting errant spaces around.  I was getting an extra space above wiremock, adding wiremock to the ordered sections resolved that so I think you'd need to add each package that's getting extra spaces above it to the config file.

**Comment by hoylemd** [comment]
I solved it!  the general problem was that somehow some local modules were appearing to import from site-packages, marking them as third party.  (probably had something to do with the egg-info directory, but I don't really know).

Solution was to add the local packages to `known_firstparty`.  For some reason it also had trouble identifying flask as a third party (it would default to first party), but I just added it to `known_thirdparty`, and it's all good.

**Comment by scorphus** [comment]
Make sure that all packages and specially Python are in the exact same version on both environments. Check #724 for reference. Please remember to close this issue once you manage to fix it.

**Comment by ahonnecke** [comment]
They seemingly are, from what I can tell.  Is there a further check that I can perform?

### In the docker container

```bash
circleci@21868dbbaf66:~/project/repos/web$ pipenv run isort --version

/#######################################################################\

     `sMMy`
     .yyyy-                                                      `
    ##soos##                                                    ./o.
          `     ``..-..`         ``...`.``         `   ```` ``-ssso```
     .s:-y-   .+osssssso/.     ./ossss+:so+:`    :+o-`/osso:+sssssssso/
     .s::y-   osss+.``.``     -ssss+-.`-ossso`   ssssso/::..::+ssss:::.
     .s::y-   /ssss+//:-.`   `ssss+     `ssss+   sssso`       :ssss`
     .s::y-   `-/+oossssso/  `ssss/      sssso   ssss/        :ssss`
     .y-/y-       ````:ssss`  ossso.    :ssss:   ssss/        :ssss.
     `/so:`    `-//::/osss+   `+ssss+-/ossso:    /sso-        `osssso/.
       \/      `-/oooo++/-      .:/++:/++/-`      ..           `://++/.


         isort your Python imports for you so you don't have to

                            VERSION 4.3.4

\########################################################################/

circleci@21868dbbaf66:~/project/repos/web$ pipenv run python --version
Python 3.6.4
```

### Locally

```bash
➜ pipenv run isort --version

/#######################################################################\

     `sMMy`
     .yyyy-                                                      `
    ##soos##                                                    ./o.
          `     ``..-..`         ``...`.``         `   ```` ``-ssso```
     .s:-y-   .+osssssso/.     ./ossss+:so+:`    :+o-`/osso:+sssssssso/
     .s::y-   osss+.``.``     -ssss+-.`-ossso`   ssssso/::..::+ssss:::.
     .s::y-   /ssss+//:-.`   `ssss+     `ssss+   sssso`       :ssss`
     .s::y-   `-/+oossssso/  `ssss/      sssso   ssss/        :ssss`
     .y-/y-       ````:ssss`  ossso.    :ssss:   ssss/        :ssss.
     `/so:`    `-//::/osss+   `+ssss+-/ossso:    /sso-        `osssso/.
       \/      `-/oooo++/-      .:/++:/++/-`      ..           `://++/.


         isort your Python imports for you so you don't have to

                            VERSION 4.3.4

\########################################################################/


web on  handle-soft-timeout [⇡$] via 🐍 3.6.4
➜ pipenv run python --version
Python 3.6.4
```

**Comment by scorphus** [comment]
One thing I can think of, is there any cache being saved and restored? That may leave the container virtualenv in a different state than your local one.

Or may I assume you're building and running the container locally? I did that with the following Dockerfile and got exact same results from both the local container's virtualenv and my local virtualenv. See:

```Dockerfile
FROM circleci/python:3.6.4

USER root

RUN mkdir /src /venv
COPY . /src/
RUN chown -R circleci:circleci /src /venv

USER circleci
WORKDIR /src

RUN python3 -m venv ../venv && \
    . ../venv/bin/activate && \
    pip install -r requirements.txt

CMD python main.py
```

From container's virtualenv:
```
(venv) circleci@850827ac06ca:/src$ python --version
Python 3.6.4
(venv) circleci@850827ac06ca:/src$ isort -c -vb

/#######################################################################\

     `sMMy`
     .yyyy-                                                      `
    ##soos##                                                    ./o.
          `     ``..-..`         ``...`.``         `   ```` ``-ssso```
     .s:-y-   .+osssssso/.     ./ossss+:so+:`    :+o-`/osso:+sssssssso/
     .s::y-   osss+.``.``     -ssss+-.`-ossso`   ssssso/::..::+ssss:::.
     .s::y-   /ssss+//:-.`   `ssss+     `ssss+   sssso`       :ssss`
     .s::y-   `-/+oossssso/  `ssss/      sssso   ssss/        :ssss`
     .y-/y-       ````:ssss`  ossso.    :ssss:   ssss/        :ssss.
     `/so:`    `-//::/osss+   `+ssss+-/ossso:    /sso-        `osssso/.
       \/      `-/oooo++/-      .:/++:/++/-`      ..           `://++/.


         isort your Python imports for you so you don't have to

                            VERSION 4.3.4

\########################################################################/

else-type place_module for json returned STDLIB
else-type place_module for logging returned STDLIB
else-type place_module for os returned STDLIB
else-type place_module for sys returned STDLIB
else-type place_module for traceback returned STDLIB
from-type place_module for collections returned STDLIB
from-type place_module for datetime returned STDLIB
from-type place_module for logging.config returned STDLIB
from-type place_module for XPTO returned FIRSTPARTY
SUCCESS: /src/somedir/somefile.py Everything Looks Good!
SUCCESS: /src/somedir/somefile.py Everything Looks Good!
SUCCESS: /src/somedir/somefile.py Everything Looks Good!
else-type place_module for datetime returned STDLIB
else-type place_module for json returned STDLIB
else-type place_module for logging returned STDLIB
else-type place_module for os returned STDLIB
else-type place_module for smtplib returned STDLIB
else-type place_module for ssl returned STDLIB
else-type place_module for time returned STDLIB
else-type place_module for uuid returned STDLIB
from-type place_module for email.mime.text returned STDLIB
from-type place_module for typing returned STDLIB
else-type place_module for etcd3 returned THIRDPARTY
else-type place_module for pytz returned THIRDPARTY
from-type place_module for jinja2 returned THIRDPARTY
from-type place_module for tornado.ioloop returned THIRDPARTY
SUCCESS: /src/somedir/somefile.py Everything Looks Good!
from-type place_module for .allocation_agent returned LOCALFOLDER
SUCCESS: /src/somedir/somefile.py Everything Looks Good!
SUCCESS: /src/somedir/somefile.py Everything Looks Good!
SUCCESS: /src/somedir/somefile.py Everything Looks Good!
else-type place_module for datetime returned STDLIB
from-type place_module for os returned STDLIB
from-type place_module for time returned STDLIB
from-type place_module for unittest returned STDLIB
from-type place_module for unittest.mock returned STDLIB
from-type place_module for xml.etree returned STDLIB
else-type place_module for pytz returned THIRDPARTY
from-type place_module for XPTO FIRSTPARTY
from-type place_module for XPTO FIRSTPARTY
SUCCESS: /src/somedir/somefile.py Everything Looks Good!
SUCCESS: /src/somedir/somefile.py Everything Looks Good!
(venv) circleci@850827ac06ca:/src$ cat .isort.cfg 
[settings]
line_length = 100
lines_after_imports = 2
not_skip =
    __init__.py
(venv) circleci@850827ac06ca:/src$
```
From local virtualenv:
```
(venv) scorphus@mac:~/repo$ python --version
Python 3.6.4
(venv) scorphus@mac:~/repo$ isort -c -vb

/#######################################################################\

     `sMMy`
     .yyyy-                                                      `
    ##soos##                                                    ./o.
          `     ``..-..`         ``...`.``         `   ```` ``-ssso```
     .s:-y-   .+osssssso/.     ./ossss+:so+:`    :+o-`/osso:+sssssssso/
     .s::y-   osss+.``.``     -ssss+-.`-ossso`   ssssso/::..::+ssss:::.
     .s::y-   /ssss+//:-.`   `ssss+     `ssss+   sssso`       :ssss`
     .s::y-   `-/+oossssso/  `ssss/      sssso   ssss/        :ssss`
     .y-/y-       ````:ssss`  ossso.    :ssss:   ssss/        :ssss.
     `/so:`    `-//::/osss+   `+ssss+-/ossso:    /sso-        `osssso/.
       \/      `-/oooo++/-      .:/++:/++/-`      ..           `://++/.


         isort your Python imports for you so you don't have to

                            VERSION 4.3.4

\########################################################################/

else-type place_module for json returned STDLIB
else-type place_module for logging returned STDLIB
else-type place_module for os returned STDLIB
else-type place_module for sys returned STDLIB
else-type place_module for traceback returned STDLIB
from-type place_module for collections returned STDLIB
from-type place_module for datetime returned STDLIB
from-type place_module for logging.config returned STDLIB
from-type place_module for tmh.controller.allocation_agent returned FIRSTPARTY
SUCCESS: /Users/pablo/Workspace/mobilityhouse/tmh-nl-tennet-allocation-agent/main.py Everything Looks Good!
SUCCESS: /Users/pablo/Workspace/mobilityhouse/tmh-nl-tennet-allocation-agent/tests/__init__.py Everything Looks Good!
SUCCESS: /Users/pablo/Workspace/mobilityhouse/tmh-nl-tennet-allocation-agent/tests/unit/placeholder.py Everything Looks Good!
else-type place_module for datetime returned STDLIB
from-type place_module for os returned STDLIB
from-type place_module for time returned STDLIB
from-type place_module for unittest returned STDLIB
from-type place_module for unittest.mock returned STDLIB
from-type place_module for xml.etree returned STDLIB
else-type place_module for pytz returned THIRDPARTY
from-type place_module for tmh.controller returned FIRSTPARTY
from-type place_module for tmh.controller.allocation_agent returned FIRSTPARTY
SUCCESS: /Users/pablo/Workspace/mobilityhouse/tmh-nl-tennet-allocation-agent/tests/unit/test_allocation_agent.py Everything Looks Good!
SUCCESS: /Users/pablo/Workspace/mobilityhouse/tmh-nl-tennet-allocation-agent/tests/func/placeholder.py Everything Looks Good!
SUCCESS: /Users/pablo/Workspace/mobilityhouse/tmh-nl-tennet-allocation-agent/tmh/__init__.py Everything Looks Good!
SUCCESS: /Users/pablo/Workspace/mobilityhouse/tmh-nl-tennet-allocation-agent/tmh/controller/__init__.py Everything Looks Good!
from-type place_module for .allocation_agent returned LOCALFOLDER
SUCCESS: /Users/pablo/Workspace/mobilityhouse/tmh-nl-tennet-allocation-agent/tmh/controller/allocation_agent/__init__.py Everything Looks Good!
else-type place_module for datetime returned STDLIB
else-type place_module for json returned STDLIB
else-type place_module for logging returned STDLIB
else-type place_module for os returned STDLIB
else-type place_module for smtplib returned STDLIB
else-type place_module for ssl returned STDLIB
else-type place_module for time returned STDLIB
else-type place_module for uuid returned STDLIB
from-type place_module for email.mime.text returned STDLIB
from-type place_module for typing returned STDLIB
else-type place_module for etcd3 returned THIRDPARTY
else-type place_module for pytz returned THIRDPARTY
from-type place_module for jinja2 returned THIRDPARTY
from-type place_module for tornado.ioloop returned THIRDPARTY
SUCCESS: /Users/pablo/Workspace/mobilityhouse/tmh-nl-tennet-allocation-agent/tmh/controller/allocation_agent/allocation_agent.py Everything Looks Good!
(venv) scorphus@mac:~/repo$ cat .isort.cfg 
[settings]
line_length = 100
lines_after_imports = 2
not_skip =
    __init__.py
(venv) scorphus@mac:~/repo$
```

Try running all from local and see where you get from there. I *had* the exact same issue, BTW.

I hope this helps you solve your issue as well.

**Comment by ahonnecke** [comment]
Not running it locally, but actually in the CI pipeline.  I'll try again with ```-c -vb``` thanks!

**Comment by ahonnecke** [comment]
[isort.version.local.txt](https://github.com/timothycrosley/isort/files/2380551/isort.version.local.txt)

**Comment by aarcro** [comment]
Having this issue, There's actually different behavior between `pipenv run isort` (wrong) and `pipenv shell` then `isort` inside the subshell (correct).

**Comment by gte620v** [comment]
in case anyone else is having issues with this, there is a typo in the suggested fix above.  `known_thirdparty` should be `known_third_party`.

**Comment by ahonnecke** [comment]
FWIW pipenv shell as opposed to pipenv run is not available in all environments (circleci)

**Comment by zwing99** [comment]
This issue is AWEFUL. We want to use this and yeah it seems like i have to over config it in order to use it in a bring your own os environment that i work it. Can someone please repair this behavior!

**Comment by timothycrosley** [comment]
I'm so sorry this issue has affected all of you! isort does a lot of work to try to intelligently determine the placement of imports - unfortunately, for all of you, it so far has proved to be too clever by half.

Is anyone willing to see if this behavior is addressed for them in the latest release: 4.3.13? I've done a ton of work on trying to ensure consistency, especially across Operating System, and dependency inclusion systems (adding conda, PipEnv, RequirementsFile, support etc.) I've also ensured that all tests are ran across Mac, Linux, and Windows. Finally, I've set up personal VMs for the two Operating Systems I don't personally use (Mac and Windows) to do as much in-depth analysis as possible. 

Thank you!

~Timothy

**Comment by ahonnecke** [comment]
I can confirm that this behaves the same way (without a config file) in OSX 10.13.04 and in a container running amazonlinux:2017.09 (centoS)

**Comment by choich** [comment]
I don't think this problem is completely fixed.
My local (mac os + pipenv) and jenkins ci's isort works differently.
It also works differently inside and outside pipenv on the same mac local.
Can you confirm this?

If I run isort outside of pipenv in local, the files are fixed.
![스크린샷 2019-09-04 오전 10 46 40](https://user-images.githubusercontent.com/13517510/64219676-5cc76880-cf01-11e9-9d35-14bb0e2f1aba.png)

And when I run isort inside the pipenv shell, the files are fixed again.
![스크린샷 2019-09-04 오전 10 47 05](https://user-images.githubusercontent.com/13517510/64219740-ac0d9900-cf01-11e9-995d-b786567be87b.png)

**Comment by coltonbh** [comment]
I'm seeing differential sorting my local MacOS if I use the system install of isort vs `pipenv run isort`. The version of isort is the same on both installs, yet I get differential sorting. 

The cause seems to be that running the system install of isort considers packages used in my module but not installed in the system python (for example the `sqlalchemy` package) as FIRSTPARTY instead of THIRDPARTY (isort thinks these modules are part of my own code package), so it groups them with my package-specific imports. When I use `pipenv run isort` it correctly identifies the installed packages as THIRDPARTY and my own package as FIRSTPARTY.

**Comment by vjpr** [comment]
I just got stung by this.

```
isort --verbose
```

On a CircleCI environment running Debian 10:

```
from-type place_module for environs returned FIRSTPARTY
```

macOS:

```
from-type place_module for environs returned THIRDPARTY
```

---

In CircleCI env I am installing using `pip install isort`, and then locally I am using `poetry run isort`.

**Comment by alexellison** [comment]
@vjpr I also just ran into this issue (using `v4.3.21` and same debian/macos envs you describe) and was able to resolve by specifying `default_section = "THIRDPARTY"` in my config.

Note: my config syntax is for a `pyproject.toml` file so you may have to use `default_section=THIRDPARTY` if you're using `.isort.cfg` or `setup.cfg`.
